src/aktualizr_lite/main.cc

```diff
@@ -239,20 +242,43 @@ static int daemon_main(LiteClient &client, const bpo::variables_map &variables_m
     }
 
     auto target = find_target(client.primary, hwid, client.config.pacman.tags, "latest");
-    if (target != nullptr && !targets_eq(*target, current, compareDockerApps)) {
-      LOG_INFO << "Updating base image to: " << *target;
-
-      data::ResultCode::Numeric rc = do_update(client, *target, lockfile);
-      if (rc == data::ResultCode::Numeric::kOk) {
-        current = *target;
-        client.http_client->updateHeader("x-ats-target", current.filename());
-        // Start the loop over to call updateImagesMeta which will update this
-        // device's target name on the server.
-        continue;
-      } else if (rc == data::ResultCode::Numeric::kNeedCompletion) {
-        if (std::system(client.config.bootloader.reboot_command.c_str()) != 0) {
-          LOG_ERROR << "Unable to reboot system";
-          return 1;
+    if (target != nullptr) {
+      // This is a workaround for finding and avoiding bad updates after a rollback.
+      // Rollback sets the installed version state to none instead of broken, so there is no
+      // easy way to find just the bad versions without api/storage changes. As a workaround we
+      // just check if the version is known (old hash) and not current/pending and abort if so
+      bool known_target_sha = false;
+      boost::optional<Uptane::Target> pending;
+      client.storage->loadInstalledVersions("", nullptr, &pending);
+      if (target->sha256Hash() != current.sha256Hash()) {
+        std::vector<Uptane::Target>::reverse_iterator it;
+        for (it = installed_versions.rbegin(); it != installed_versions.rend(); it++) {
+          if (it->sha256Hash() == target->sha256Hash()) {
+            // Make sure installed version is not what is currently pending
+            if ((pending != boost::none) && (it->sha256Hash() == pending->sha256Hash())) {
+              continue;
+            }
+            LOG_INFO << "Target sha256Hash " << target->sha256Hash() << " known locally (rollback?), skipping";
+            known_target_sha = true;
```

Comment:
> This might be problem for an edge case. Lets say we have two builds, 1 and 2 both tagged "promoted" and the device listens only accepts promoted. If you removed the "promoted" tag from build 2 and update for build 1 would get triggered here. I _think_ this logic would reject the update, right? There is a `was_installed` parameter to the `loadInstalledVersions` method. I wonder if we could use that to get a list of failed installs?

The problem is that every image state gets stored into installed versions, like after applying the install (pending), after confirming the boot (current, when was_installed is also set to 1) and after marking the image as bad (none). If we are able to have a separated field for 'broken' images, then we could use the install log to only retrieve the ones we want, as currently we can only retrieve ones that were set to current || everything.

Reply:
What we could do instead is using was_installed to retrieve the list of good images and assume it is not bad if part of the list or equal to pending, but that would basically be a variation of what this piece is already doing.